README.md
MOOG_SCAT
==========

MOOG_SCAT .

Authors
-------

- **Jennifer Sobeck** (CFHT/UW)
- **Chris Sneden** (UT Austin)

Installation
------------

MOOG_SCAT is a Fortran 77 (F77) code.  To install the latest version of the code, employ one of the OS-specific Makefiles (Makefile_SCAT.macos; Makefile_SCAT.rh) and 
use the following command:

    make -f Makefile*

Accordingly, a Fortran compiler is necessary (succesful MOOG_SCAT compilation has been achieved with a ``gfortran`` compiler).  In addition, MOOG_SCAT relies upon 
SuperMongo for plotting/visualziation.	
    
Documentation
-------------

Refer to `the documentation <http:https://moog-scat.readthedocs.io/en/latest///>`_ for information on how
to install and use MOOG_SCAT.


docs/changelog.rst
MOOG_SCAT v1.0.0 CHANGELOG
==========================

Purpose:

- update all LTE subroutines so that they are consistent with the current 2019 release

- general update/vet of MOOG-SCAT code in order to issue new release

General log of changes:

- Compared MOOG_LTE v.2019 to MOOG_SCAT v.2017-2020 (unreleased)

- Adopted and integrated all non-SCAT MOOG-LTE subroutines

- Modified, as needed, SCAT-associated routines to be consistent with their LTE counterparts

- Uncovered a few errors in MOOG-LTE v.2019 (release Nov 2019) and altered them in current SCAT code version

List of LTE Coverted Routines
-----------------------------

(All MOOG Drivers; ! indicates unxpected version differences)

Abfind --> Abfind_SCAT

Abpop --> Abpop_SCAT (!; resolved)

Binary --> Binary_SCAT

Blends --> Blends_SCAT (!; resolved)

Cdcalc --> Cdcalc_SCAT

Cog --> Cog_SCAT

Cogsyn --> Cogsyn_SCAT

Curve --> Curve_SCAT

Doflux --> Doflux_SCAT

Ewfind --> Ewfind_SCAT (!)

Fakeline --> Fakeline_SCAT

Getsyns --> Getsyns_SCAT

Gridsyn --> Gridsyn_SCAT

Lineabund --> Lineabund_SCAT

Various Makefiles (!)

Moog --> Moog_SCAT (!; resolved)

Moogsilent -->Moogsilent_SCAT (!; resolved)

Oneline --> Oneline_SCAT (!; resolved??---> check equations/theory)

Synpop --> Synpop_SCAT (!; resolved)

Synspec --> Synspec_SCAT(!; resolved)

Synth --> Synth_SCAT

SCAT-Only Routines/Files
------------------------

Scat.com

Ang_Weight_SCAT.f

Gauss_Quad_Sum.f

Sourcefunc_cont_SCAT.f

Sourcefunc_line_SCAT.f

LTE Code Changes
----------------

- Removal of the ESTIM routine (as compared to earlier LTE [2017, 2014] versions)

- Removal of the Minimax routine (as compared to earlier LTE [2017, 2014] versions)

- Addition of SETMOLS routine (as compared to earlier LTE [2017, 2014] versions)

- In Moog_silent.f, there is still redundancy if a user calls the deprecated 'isotop' driver while in Moog.f no such redundancy exists. Consistency issue!

JS Suggested MOOG_LTE Modifications
-----------------------------------

- Removal of PLOTVALS common block (still mentioned in Correl.f; but as Correl.f subroutine does not employ any of the VAR, should be fine)

- Modification of Correl.f to remove PLOTVALS common block

- Removal of common block file from distribution

JS To Do
--------

- Modified LTE routines list

- New/additional SCAT routines

- How to call MOOG_SCAT, e.g., ./MOOG_SCAT

- General Documentation

LTE - SCAT Common Subroutines: Difference (dff) Comparisons
-----------------------------------------------------------
>>> Abfind
2c2
<       subroutine abfind
---
>       subroutine abfind_scat
60c60
< 100   call fakeline
---
> 100   call fakeline_scat
92c92
<             call lineabund (abundin)
---
>             call lineabund_scat (abundin)
105c105
<             call lineabund (abundin)
---
>             call lineabund_scat (abundin)

>>> Abpop [!ONE CHANGE; RESOLVED]
2c2
<       subroutine abpop
---
>       subroutine abpop_scat
101c101
<             call curve
---
>             call curve_scat
223d222
<       if (silent .eq. 'n') then 
227,229d225
<       else
<          choice = 'y'
<       endif

>>>Binary
c2
<       subroutine binary
---
>       subroutine binary_scat
111c111
<          call synspec
---
>          call synspec_scat
122c122
<             call synspec
---
>             call synspec_scat

>>> Blends [!UNEXPECTED CHANGES]
2c2
<       subroutine blends 
---
>       subroutine blends_scat
14a15
>       include 'Scat.com'
122c123
<             call synspec
---
>             call synspec_scat
180c181
<          call synspec
---
>          call synspec_scat
194,196c195,200
<             call cdcalc(2)
<             first = 0.4343*cd(1)
<             d(1) = rinteg(xref,cd,dummy1,ntau,first)
---
> c            call cdcalc (2)
>             call cdcalc_scat(2)
> c            first = 0.4343*cd(1)
>             first = 0.4343*adepth
> c            d(1) = rinteg(xref,cd,dummy1,ntau,first)
>             d(1) = adepth
198c202,203
<                dummy1(i) = xref(i)*cd(i)
---
> c               dummy1(i) = xref(i)*cd(i)
>                dummy1(i) = xref(i)*adepth
200a206,207
> c            cdmean = rinteg(xref,dummy1,dummy2,ntau,first)/
> c     .      rinteg(xref,cd,dummy2,ntau,first)
202c209
<      .      rinteg(xref,cd,dummy2,ntau,first)
---
>      .      rinteg(xref,adepth,dummy2,ntau,first)
204c211,212
<                if (cdmean .lt. cd(i)) exit
---
> c               if (cdmean .lt. cd(i)) exit
>                if (cdmean .lt. adepth) exit


>>> Cdcalc 

>>> Cog
2c2
<       subroutine cog
---
>       subroutine cog_scat
73c73
<          call curve
---
>          call curve_scat

>>> Cogsyn
<       subroutine cogsyn
---
>       subroutine cogsyn_scat
68c68
<       call synspec
---
>       call synspec_scat
83c83
<       call synspec
---
>       call synspec_scat


>>> Curve
2c2
<       subroutine curve
---
>       subroutine curve_scat
24c24
< 31    call oneline (2)
---
> 31    call oneline_scat (2)
41c41
< 61    call oneline (2)
---
> 61    call oneline_scat (2)


>>> Doflux
2c2
<       subroutine doflux
---
>       subroutine doflux_scat
11a12
>       include 'Scat.com'
53,55c54,58
<       call cdcalc (1)
<       first = 0.4343*cd(1)
<       flux = rinteg(xref,cd,dummy1,ntau,first)
---
> c      call cdcalc (1)
>       call cdcalc_scat (1)
> c      first = 0.4343*cd(1)
> c      flux = rinteg(xref,cd,dummy1,ntau,first)
>       flux = Flux_cont_MOOG

>>> EWfind [?UNEXPECTED CHANGES]
,2c1
< 
<       subroutine ewfind
---
>       subroutine ewfind_scat
12a12
>       include 'Scat.com'
16a17
> 
19a21
> 
51a54
>       isynth = 1
69,70c72,73
<          array(1:40) = 'wavelength        EP     logGF     ident'
<          array(41:60) = '     Abund    EWcalc'
---
>          call molquery
>          write (array,1001)
77c80
<          call oneline (1)
---
>          call oneline_scat (1)
79a83,85
> 
> 
>          if (iatom .lt. 100) then
80a87,89
>          else
>             xab = dlog10(xabund(iabtom)) + 12.

Begin.f

      subroutine begin
c***************************************************************************
c     This routine simply starts up MOOG
c     THIS VERSION IS FOR LINUX REDHAT MACHINES
c***************************************************************************

      implicit real*8 (a-h,o-z)
      include 'Atmos.com'
      include 'Pstuff.com'
      character*80 line, systemcall
      integer num


c*****define the number of text screen lines for silent mode;
c     this number is hardwired, since it is not really needed at run time.
      if (silent .eq. 'y') then
         maxline = 24
           write (*,*) 'maxline', maxline
           pause
         go to 10
      endif


c*****define the number of lines available on the text screen for
c     interactive mode; this number is discovered from the "stty -a"
c     command, for which the output format is unique to the operating
c     system.
      write (systemcall,*) 'stty -a > tmpsize'
      call system (systemcall)
      open (99,file='tmpsize')
5     read (99,1010,end=15) line
      do i=1,77
         if (line(i:i+3) .eq. 'rows') then
            if     (machine .eq. 'pcl') then
            read (line(i+4:i+6),1011) maxline
            elseif (machine .eq. 'mac') then
               read (line(i-4:i-2),1011) maxline
            elseif (machine .eq. 'uni') then
               read (line(i+6:i+8),1011) maxline
            endif
            go to 10
         endif
      enddo
      go to 5
15    array = 'SCREEN ROW COUNT UNKNOWN; USE 24 ([y]/n)? '
      nchars = 42
      ikount = 2
      call getasci (nchars,ikount)
      choice = chinfo(1:1)
      if (choice.eq.'y' .or. nchars.le.0) then
         go to 10
      else
         call finish (0)
      endif
10    close (99,status='delete')
      write (systemcall,*) '\\rm -f tmpsize'
      call system (systemcall)
      if (maxline .lt. 10) then
         maxline = 24
      else
         maxline = maxline - 2
      endif


c*****clear the text screen
      write (systemcall,*) 'clear'
      call system (systemcall)


c*****open data files carried with the source code: Barklem damping
      nfbarklem = 35
      num = 60
      call getcount (num,moogpath)
      if (moogpath(num:num) .ne. '/') then
         num = num + 1
         moogpath(num:num) = '/'
      endif
      fbarklem(1:num) = moogpath(1:num)
      fbarklem(num+1:num+11) = 'Barklem.dat'
      open (nfbarklem,file=fbarklem)


c*****open data files carried with the source code: Barklem UV damping
      nfbarklemUV = 36
      num = 60
      call getcount (num,moogpath)
      if (moogpath(num:num) .ne. '/') then
         num = num + 1
         moogpath(num:num) = '/'
      endif
      fbarklemUV(1:num) = moogpath(1:num)
      fbarklemUV(num+1:num+13) = 'BarklemUV.dat'
      open (nfbarklemUV,file=fbarklemUV)
 

c  write a header and find the appropriate parameter file, and exit normally
      write (array,1001)
      istat = ivwrite (1,1,array,79)
      write (array,1004)
      istat = ivwrite (2,1,array,79)
      array = 'MOOG PARAMETERS? ' 
      nchars = 15
      nfparam = 50     
      lscreen = 4
      if (silent .eq. 'y') then
         fparam = 'batch.par'
      else
         fparam = 'no_filename_given'     
      endif
      call infile ('input  ',nfparam,'formatted  ',0,nchars,
     .             fparam,lscreen)
      read (nfparam,1002) control
      write (array,1003) control
      istat = ivwrite (2,1,array,58)
      write (array,1001)
      istat = ivwrite (3,1,array,79)
      return


c*****format statements
1001  format (79('*'))
1002  format (a7)
1003  format (22x,'MOOG IS CONTROLLED BY DRIVER ',a7)
1004  format (25(' '),'MOOG SCAT VERSION (SEP 2022)',26(' '))   
1010  format (a80)
1011  format (i3)


      end
      









Moog_SCAT.f

      program moog_scat
c******************************************************************************
c     This is the main driver for MOOG.  It reads the parameter
c     file and sends MOOG to various controlling subroutines.
c     This is the normal interactive version of the code; for batch
c     processing without user decisions, run MOOGSILENT instead.
c******************************************************************************

      include 'Atmos.com'
      include 'Pstuff.com'
      character yesno*1


c$$$$$$$$$$$$$$$$$$$$$$$$ USER SETUP AREA $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
c*****in compiling MOOG, here the various machine-specific things are 
c     declared.  First, define the directory where MOOG lives, in order to 
c     be able to pull in auxiliary data files; executing 'make' will 
c     generate a reminder of this necessity
      write (moogpath,1001)
      moogpath = 
     .  '/Users/jsobeck/moog_SCAT-update'


c*****What kind of machine are you using?  Possible ones are:
c     "mac" = Intel-based Apple Mac 
c     "pcl" = a PC or desktop running some standard linux like Redhat
c     "uni" = a machine running Unix, specifically Sun Solaris
      machine = "mac"


c*****for x11 terminal types, define the parameters of plotting windows;
c     set up an x11 screen geometry and placement that is good for spectrum
c     syntheses (long, but not tall); the user should play with the format
c     statements for particular machines.
      write (smt1,1018)
c     now do the same for line abundance trend plots (short but tall).
      write (smt2,1017)
c$$$$$$$$$$$$$$$$$$$$$$$ END OF USER SETUP $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$


c*****declare this to be the normal interactive version; variable "silent"
c     will be queried on all occasions that might call for user input;
c     DON'T CHANGE THIS VARIABLE; 
c     if silent = 'n', the normal interactive MOOG is run;
c     if silent = 'y', the non-interactive MOOG is run
      silent = 'n'


c*****invoke the overall starting routine
1     control = '       '
      call begin

c*****Call the Ang_Weight_SCAT routine
	call ang_weight_SCAT

c*****use one of the standard driver routines
      if     (control .eq. 'synplot') then
         call plotit
      elseif (control .eq. 'synth  ') then
         call synth_scat
      elseif (control .eq. 'cogsyn ') then
         call cogsyn_scat  
      elseif (control .eq. 'blends ') then
         call blends_scat  
      elseif (control .eq. 'abfind ') then
         call abfind_scat
      elseif (control .eq. 'ewfind ') then
         call ewfind_scat
      elseif (control .eq. 'cog    ') then
         call cog_scat
      elseif (control .eq. 'calmod ') then
         call calmod
      elseif (control .eq. 'doflux ') then
         call doflux_scat   
      elseif (control .eq. 'weedout') then
         call weedout  
      elseif (control .eq. 'gridsyn') then
         call gridsyn_scat  
      elseif (control .eq. 'gridplo') then
         call gridplo  
      elseif (control .eq. 'binary ') then
         call binary_scat
      elseif (control .eq. 'abpop  ') then
         call abpop_scat
      elseif (control .eq. 'synpop ') then
         call synpop_scat


c*****or, put in your own drivers in the form below....
      elseif (control .eq. 'mine  ') then
         call  mydriver 


c*****or else you are out of luck!
      else
         array = 'THIS IS NOT ONE OF THE DRIVERS.  TRY AGAIN (y/n)?'
         istat = ivwrite (4,3,array,49)
         istat = ivmove (3,1)
         read (*,*) yesno
         if (yesno .eq. 'y') then
            go to 1
         else
            call finish (0)
         endif
      endif

c*****format statements
1001  format (60(' '))
1003  format (22x,'MOOG IS CONTROLLED BY DRIVER ',a7)
1017  format ('x11 -bg black -title MOOGplot -geom 700x800+650+000')
1018  format ('x11 -bg black -title MOOGplot -geom 1200x350+20+450')


      end


      


Moogsilent_SCAT.f

      program moogsilent_scat
c******************************************************************************
c     This is the main driver for the "non-interactive" version of MOOG.
c     It reads the parameter file and sends MOOG to various controlling
c     subroutines.  In this version of MOOG, the parameter file must
c     be named "batch.par" (because the code cannot stop to ask the
c     user to name the parameter file).      
c******************************************************************************

      include 'Atmos.com'
      include 'Pstuff.com'


c$$$$$$$$$$$$$$$$$$$$$$$$ USER SETUP AREA $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
c     In compiling MOOG, here the various machine-specific things are 
c     declared.  First, define the directory where MOOG lives, in order to 
c     be able to pull in auxiliary data files; executing 'make' will 
c     generate a reminder of this
      write (moogpath,1001)
      moogpath = 
     .  '/Users/jsobeck/moog_SCAT-update'


c*****What kind of machine are you using?  Possible ones are:
c     "mac" = Intel-based Apple Mac 
c     "pcl" = a PC or desktop running some standard linux like Redhat
c     "uni" = a machine running Unix, specifically Sun Solaris
      machine = "mac"


c*****for x11 terminal types, define the parameters of plotting windows;
c     set up an x11 screen geometry and placement that is good for spectrum
c     syntheses (long, but not tall); the user should play with the format
c     statements for particular machines.
      write (smt1,1018)
c     now do the same for line abundance trend plots (short but tall).
      write (smt2,1017)
c$$$$$$$$$$$$$$$$$$$$$$$ END OF USER SETUP $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$


c*****declare this to be the non-interactive version; variable "silent"
c     will be queried on all occasions that might call for user input;
c     DON'T CHANGE THIS VARIABLE; 
c     if silent = 'n', the normal interactive MOOG is run;
c     if silent = 'y', the non-interactive MOOG is run
      silent = 'y'


c*****invoke the overall starting routine
      control = '       '
      call begin


c*****use one of the standard driver routines ("isotop" is obsolete):
      if     (control .eq. 'synplot') then
         call plotit
      elseif (control .eq. 'isoplot') then
         call plotit         
      elseif (control .eq. 'synth  ') then
         call synth_scat
      elseif (control .eq. 'cogsyn ') then
         call cogsyn_scat  
      elseif (control .eq. 'blends ') then
         call blends_scat  
      elseif (control .eq. 'abfind ') then
         call abfind_scat
      elseif (control .eq. 'ewfind ') then
         call ewfind_scat
      elseif (control .eq. 'cog    ') then
         call cog_scat
      elseif (control .eq. 'calmod ') then         
         call calmod
      elseif (control .eq. 'isotop ') then
         control = 'synth  '
         call synth_scat   
      elseif (control .eq. 'doflux ') then
         call doflux_scat   
      elseif (control .eq. 'weedout') then
         call weedout  
      elseif (control .eq. 'gridsyn') then
         call gridsyn_scat  
      elseif (control .eq. 'gridplo') then
         call gridplo  
      elseif (control .eq. 'binary ') then
         call binary_scat
      elseif (control .eq. 'abpop  ') then
         call abpop_scat
      elseif (control .eq. 'synpop ') then
         call synpop_scat


c*****or, put in your own drivers in the form below....
      elseif (control .eq. 'mine  ') then
         call  mydriver 


c*****or else you are out of luck!
      else
         array = 'THIS IS NOT ONE OF THE DRIVERS. I QUIT!'
         istat = ivwrite (4,3,array,49)
         stop
      endif


c*****Call the Ang_Weight_SCAT routine
	call ang_weight_SCAT

c*****format statements
1001  format (60(' '))
1017  format ('x11 -bg black -title MOOGplot -geom 700x800+650+000')
1018  format ('x11 -bg black -title MOOGplot -geom 1200x400+20+450')


      end


      
